pkg/rt/manager.go

```diff
@@ -68,32 +72,48 @@ func (m *Manager) Refresh() {
 	defer m.mtx.Unlock()
 
 	nextgen := map[string]interrupt{}
-	for _, id := range m.ids.ServiceIDs() {
-		if irq, ok := m.managed[id]; ok {
-			level.Debug(m.logger).Log("service_id", id, "subscriber", "maintain")
-			nextgen[id] = irq // move
-			delete(m.managed, id)
-		} else {
-			level.Info(m.logger).Log("service_id", id, "subscriber", "create")
-			nextgen[id] = m.spawn(id)
+	for _, product := range api.Products {
+		if m.productCache.HasAccess(product) {
+			for _, id := range m.ids.ServiceIDs() {
+				key := id + sep + product
+
+				if irq, ok := m.managed[key]; ok {
+					level.Debug(m.logger).Log("service_id", id, "type", product, "subscriber", "maintain")
+					nextgen[key] = irq // move
+					delete(m.managed, key)
+				} else {
+					level.Info(m.logger).Log("service_id", id, "type", product, "subscriber", "create")
+					nextgen[key] = m.spawn(id, product)
+				}
+			}
 		}
-	}
 
-	for _, id := range m.managedIDsWithLock() {
-		level.Info(m.logger).Log("service_id", id, "subscriber", "stop")
-		irq := m.managed[id]
-		irq.cancel()
-		err := <-irq.done
-		delete(m.managed, id)
-		level.Debug(m.logger).Log("service_id", id, "interrupt", err)
-	}
+		for _, key := range m.managedIDsWithLock() {
+			id, managedProduct := parseKey(key)
+			if managedProduct != product {
+				continue
+			}
+
+			level.Info(m.logger).Log("service_id", id, "type", product, "subscriber", "stop")
```

Comment:
The Log method of the logger interface requires an even number of parameters: always key/val pairs.